---
title: "snATACseq ArchR: Cluster IDing - `r REGION`"
author: "Darren Cameron"
date: "`r format(Sys.time(), '%d/%m/%y')`"
output: 
  html_document:
    theme: paper
---



```{r packages, echo=FALSE, include=FALSE}
library(scater)
library(cowplot)
library(knitr)
```


snATACseq analysis with ArchR - cluster IDs using marker genes on gene expression matrix.

*** 

### Results {.tabset}

***

#### Gene 'Exp' {.tabset}

```{R, GeneExp_heatmap, echo=FALSE, fig.height = 10, fig.width = 10}
for (FDR_THRESH in c(MARKER_FDR)) {

 geneExp_plot <- get(paste0('geneExp_plot_FDR_', FDR_THRESH, '_Log2FC_', MARKER_LOG2FC))
 print(geneExp_plot)
 
}
```

***

#### Cluster plots

```{R, Clusters, echo=FALSE, fig.height = 10, fig.width = 10}
print(group_plot)
```

***

#### Dotplot

```{R, Avg_expr, echo=FALSE, fig.height = 10, fig.width = 10}
print(av_exp_plot)
```

***

#### UMAPs

```{R, UMAP_per_gene, echo=FALSE, fig.height = 10, fig.width = 10}
print(do.call(cowplot::plot_grid, c(list(ncol = 3), all_genes_UMAP)))
```

***

#### Bar plots

```{R, bar_plots, echo=FALSE, fig.height = 10, fig.width = 10}
print(bar_plots)
```

***

#### Browser tracks {.tabset}

***

##### DLX1

```{R, browser_DLX1, echo=FALSE, fig.height = 10, fig.width = 10}
grid::grid.newpage()
grid::grid.draw(browser_track$DLX1)
```

##### ZEB2

***

```{R, browser_ZEB2, echo=FALSE, fig.height = 10, fig.width = 10}
grid::grid.newpage()
grid::grid.draw(browser_track$ZEB2)
```

***
